validator/src/main/java/org/ethereum/beacon/validator/proposer/BeaconChainProposerImpl.java

```diff
@@ -0,0 +1,177 @@
+package org.ethereum.beacon.validator.proposer;
+
+import static java.util.Collections.emptyList;
+import static org.ethereum.beacon.core.spec.SignatureDomains.PROPOSAL;
+import static org.ethereum.beacon.core.spec.SignatureDomains.RANDAO;
+
+import java.util.List;
+import java.util.Optional;
+import org.ethereum.beacon.chain.observer.ObservableBeaconState;
+import org.ethereum.beacon.chain.observer.PendingOperations;
+import org.ethereum.beacon.consensus.SpecHelpers;
+import org.ethereum.beacon.consensus.StateTransition;
+import org.ethereum.beacon.core.BeaconBlock;
+import org.ethereum.beacon.core.BeaconBlock.Builder;
+import org.ethereum.beacon.core.BeaconBlockBody;
+import org.ethereum.beacon.core.BeaconState;
+import org.ethereum.beacon.core.operations.Attestation;
+import org.ethereum.beacon.core.operations.CasperSlashing;
+import org.ethereum.beacon.core.operations.Deposit;
+import org.ethereum.beacon.core.operations.Exit;
+import org.ethereum.beacon.core.operations.ProposerSlashing;
+import org.ethereum.beacon.core.operations.slashing.ProposalSignedData;
+import org.ethereum.beacon.core.spec.ChainSpec;
+import org.ethereum.beacon.core.state.Eth1Data;
+import org.ethereum.beacon.core.state.Eth1DataVote;
+import org.ethereum.beacon.core.state.ValidatorRecord;
+import org.ethereum.beacon.pow.DepositContract;
+import org.ethereum.beacon.validator.BeaconChainProposer;
+import org.ethereum.beacon.validator.crypto.MessageSigner;
+import tech.pegasys.artemis.ethereum.core.Hash32;
+import tech.pegasys.artemis.util.bytes.Bytes32;
+import tech.pegasys.artemis.util.bytes.Bytes8;
+import tech.pegasys.artemis.util.bytes.Bytes96;
+import tech.pegasys.artemis.util.uint.UInt24;
+import tech.pegasys.artemis.util.uint.UInt64;
+
+public class BeaconChainProposerImpl implements BeaconChainProposer {
+
+  private SpecHelpers specHelpers;
+  private ChainSpec chainSpec;
+  private StateTransition<BeaconState> blockTransition;
+  private DepositContract depositContract;
+
+  public BeaconChainProposerImpl(
+      SpecHelpers specHelpers,
+      ChainSpec chainSpec,
```

Comment:
BTW I've added `ChainSpec` as a member to `SpecHelpers`. 
I think it make sense to unite them into a single essence and not passing them both to every instance

Reply:
Or even unite 3 things: `ChainSpec`, `SpecHelpers` and transitions, since they are all represent the Spec2.0

Reply:
I don't think there is a big benefit in mixing transitions in. But definitely makes sense for `ChainSpec` and `SpecHelper`.